Fixed User Ruleset from Leaderboard

Ianlucht opened the pull request:
Made a user profile open to the desired ruleset from a beatmap rather than the profile ruleset at all times. 

Resolves #31628

Ianlucht committed:
added functionality

Ianlucht committed:
Fixed impropper logic

bdach commented on osu.Game/Users/Drawables/ClickableAvatar.cs:
```diff
@@ -33,14 +36,17 @@ public partial class ClickableAvatar : OsuClickableContainer, IHasCustomTooltip<
+        /// <param name="ruleset">The ruleset. A null value will use the profile ruleset for when the profile is displayed.</param>
         /// <param name="showCardOnHover">If set to true, the <see cref="UserGridPanel"/> will be shown for the tooltip</param>
-        public ClickableAvatar(APIUser? user = null, bool showCardOnHover = false)
+        public ClickableAvatar(APIUser? user = null, bool showCardOnHover = false, IRulesetInfo? ruleset = null)
```
I have pretty negative feelings about jamming `IRulesetInfo` into the constructor of this for two usage sites out of however many.

In my book, the leaderboard usages should locally modify `Action` to pass the ruleset.

Ianlucht replied:
I believe it is a grand total of 5 uses, and of that I am pretty sure that two more are these leader boards, one of them is a test, and then one of them is `DailyChallengeEventFeed` I think that this is a reasonable thing to do because nearly half of the uses of this constructor benefit from this change, additionally, in the daily challenge feed I think something similar can be implemented to open the profile to the daily challenge ruleset. If you would like I can update my changes to include the daily challenge and update the test to include ruleset if appropriate.

Ianlucht replied:
However, looking at the test, it does not seem to have any tests related to actually clicking on the avatar only hovering, additionally the ruleset functionality is tested under the 'TestSceneUserProfileOverlay' tests.

Ianlucht committed:
Update osu.Game/Users/Drawables/ClickableAvatar.cs agatemosu's review change request

Co-authored-by: Agatem <agaatem@outlook.com>

Ianlucht committed:
Merge branch 'master' into fix-user-ruleset-from-leaderboard

Ianlucht committed:
Revert "Update osu.Game/Users/Drawables/ClickableAvatar.cs agatemosu's review change request"

This reverts commit 82c970455043ca304c33d3c5b116a675445c4da6.

Ianlucht committed:
Reapply "Update osu.Game/Users/Drawables/ClickableAvatar.cs agatemosu's review change request"

This reverts commit b0ce6b6f90360bd894883da3141b53f141f9056a.

Ianlucht committed:
Change action rather than use constructor

Ianlucht committed:
reverted constructor

Ianlucht committed:
removed xml for ruleset

Ianlucht committed:
Update ClickableAvatar.cs

Ianlucht committed:
cleaned up using rulesets

Ianlucht commented:
Alright, I made the changes, but now there is a dependency issue in , and I have no idea how to fix this issue.  ` osu.Framework.Allocation.DependencyNotRegisteredException: The type LeaderboardScore has a dependency on OsuGame, but the dependency is not registered.` I had to add the `private OsuGame = game` there to be able to get the correct user profile up with the ruleset, but in adding that it seems to have broken the dependencies there. I don't know how to properly register it. I thought I just had to add it the way it was added to `ClickableAvatar`, but that seems incorrect.

bdach commented on osu.Game/Online/Leaderboards/LeaderboardScore.cs:
```diff
@@ -87,6 +87,9 @@ public partial class LeaderboardScore : OsuClickableContainer, IHasContextMenu,
         private ScoreManager scoreManager { get; set; } = null!;
 
+        [Resolved]
+        private OsuGame game { get; set; }
```
```suggestion
        private OsuGame? game { get; set; }
```

to fix your issue. if a resolved field is annotated as nullable, it won't throw even if the dependency isn't registered

Ianlucht replied:
Thanks I was trying that earlier and I was seeing some weird warning locally, but if it works on remote then that is problem solved.

Ianlucht committed:
Update osu.Game/Online/Leaderboards/LeaderboardScore.cs add nullable

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

Ianlucht committed:
add nullable enable for nullable OsuGame reference

canBeNull = true did not work.

Ianlucht committed:
attempted to fix formatting

Formatting error might be because of the preprocessor directive being inside the class.

Ianlucht commented:
Trying to figure out a way to have the nullable OsuGame reference where the other references are explicitly un-nullable

Ianlucht committed:
fixed formatting of nullable enable

bdach reviewed (changes requested):
Please squash and force-push all of the extra commits out of here, it's noise in history that doesn't need to be there.

Upon closer inspection this should probably also be applied to clickable avatars on the results screen:

![image](https://github.com/user-attachments/assets/4e127dcb-e65d-46d1-a539-3fea08d23ea8)

which is going to be a bit annoying to achieve and also makes me unsure about whether this change is worth it in general.

bdach commented on osu.Game/Online/Leaderboards/LeaderboardScore.cs:
```diff
@@ -87,6 +87,12 @@ public partial class LeaderboardScore : OsuClickableContainer, IHasContextMenu,
         [Resolved]
         private ScoreManager scoreManager { get; set; } = null!;
 
+        // using nullable enable here because canBeNull is not respected for OsuGame
+#nullable enable
+
+        [Resolved]
+        private OsuGame? game { get; set; }
```
That's not how this should be done. Old classes where NRT is still not on are a bit of an edge case.

```diff
diff --git a/osu.Game/Online/Leaderboards/LeaderboardScore.cs b/osu.Game/Online/Leaderboards/LeaderboardScore.cs
index 2f8e7bb7b9..438a0e2268 100644
--- a/osu.Game/Online/Leaderboards/LeaderboardScore.cs
+++ b/osu.Game/Online/Leaderboards/LeaderboardScore.cs
@@ -6,6 +6,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using JetBrains.Annotations;
 using osu.Framework.Allocation;
 using osu.Framework.Extensions.Color4Extensions;
 using osu.Framework.Graphics;
@@ -87,11 +88,9 @@ public partial class LeaderboardScore : OsuClickableContainer, IHasContextMenu,
         [Resolved]
         private ScoreManager scoreManager { get; set; } = null!;
 
-        // using nullable enable here because canBeNull is not respected for OsuGame
-#nullable enable
-
-        [Resolved]
-        private OsuGame? game { get; set; }
+        [Resolved(CanBeNull = true)]
+        [CanBeNull]
+        private OsuGame game { get; set; }
 
         public LeaderboardScore(ScoreInfo score, int? rank, bool isOnlineScope = true, bool highlightFriend = true)
         {

```
